File: python_problems/UBH5.py
from s2mpjlib import *
class  UBH5(CUTEst_problem):

#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
# 
# 
#    Problem : UBH5
#    *********
# 
#    The problem is to minimize the integral of the control magnitude needed
#    to bring a vehicle, from given position and velocity, to the origin with
#    zero velocity in a fixed amount of time.  The controls are the components
#    of the vehicle acceleration. The discretization uses the trapezoidal rule.
#    This version of the problem is a variant of UBH1, where the cumulative
#    value of the objective is maintained as an additional state variable.
# 
#    The problem is convex.
# 
#    Source: unscaled problem 5 
#    (ODE = 1, CLS = 2, GRD = 1, MET = T, SEED = 0.) in
#    J.T. Betts and W.P. Huffman,
#    "Sparse Nonlinear Programming Test Problems (Release 1.0)",
#    Boeing Computer services, Seattle, July 1993.
# 
#    SIF input: Ph.L. Toint, October 1993.
# 
#    classification = "C-CLQR2-MN-V-V"
# 
#    Number of grid points
# 
#           Alternative values for the SIF file parameters:
# IE N                   10             $-PARAMETER n=100, m=70    original value
# IE N                   100            $-PARAMETER n=1000, m=700
# IE N                   500            $-PARAMETER n=5000, m=3500
# 
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#   Translated to Python by S2MPJ version 31 X 2025
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    name = 'UBH5'

    def __init__(self, *args): 
        import numpy as np
        from scipy.sparse import csr_matrix
        nargin   = len(args)

        #%%%%%%%%%%%%%%%%%%%  PREAMBLE %%%%%%%%%%%%%%%%%%%%
        v_  = {}
        ix_ = {}
        ig_ = {}
        if nargin<1:
            v_['N'] = int(10);  #  SIF file default value
        else:
            v_['N'] = int(args[0])
# IE N                   1000           $-PARAMETER n=10000, m=7000
# IE N                   2000           $-PARAMETER n=20000, m=14000
        v_['T0'] = 0.0
        v_['TF'] = 1000.0
        v_['RN'] = float(v_['N'])
        v_['TTIME'] = v_['TF']-v_['T0']
        v_['K'] = v_['TTIME']/v_['RN']
        v_['-K/2'] = -0.5*v_['K']
        v_['0'] = 0
        v_['1'] = 1
        v_['2'] = 2
        v_['3'] = 3
        v_['4'] = 4
        v_['5'] = 5
        v_['6'] = 6
        v_['7'] = 7
        #%%%%%%%%%%%%%%%%%%%  VARIABLES %%%%%%%%%%%%%%%%%%%%
        self.xnames = np.array([])
        self.xscale = np.array([])
        intvars   = np.array([])
        binvars   = np.array([])
        irA          = np.array([],dtype=int)
        icA          = np.array([],dtype=int)
        valA         = np.array([],dtype=float)
        for I in range(int(v_['1']),int(v_['7'])+1):
            for T in range(int(v_['0']),int(v_['N'])+1):
                [iv,ix_,_] = s2mpj_ii('Y'+str(I)+','+str(T),ix_)
                self.xnames=arrset(self.xnames,iv,'Y'+str(I)+','+str(T))
        for I in range(int(v_['1']),int(v_['3'])+1):
            for T in range(int(v_['0']),int(v_['N'])+1):
                [iv,ix_,_] = s2mpj_ii('U'+str(I)+','+str(T),ix_)
                self.xnames=arrset(self.xnames,iv,'U'+str(I)+','+str(T))
        #%%%%%%%%%%%%%%%%%%  DATA GROUPS %%%%%%%%%%%%%%%%%%%
        self.gscale  = np.array([])
        self.grnames = np.array([])
        cnames       = np.array([])
        self.cnames  = np.array([])
        gtype        = np.array([])
        [ig,ig_,_] = s2mpj_ii('OBJ',ig_)
        gtype = arrset(gtype,ig,'<>')
        irA  = np.append(irA,[ig])
        icA  = np.append(icA,[ix_['Y'+str(int(v_['7']))+','+str(int(v_['N']))]])
        valA = np.append(valA,float(1.0))
        for I in range(int(v_['1']),int(v_['3'])+1):
            v_['I+3'] = 3+I
            for T in range(int(v_['1']),int(v_['N'])+1):
                v_['T-1'] = -1+T
                [ig,ig_,_] = s2mpj_ii('S'+str(I)+','+str(T),ig_)
                gtype = arrset(gtype,ig,'==')
                cnames = arrset(cnames,ig,'S'+str(I)+','+str(T))
                irA  = np.append(irA,[ig])
                icA  = np.append(icA,[ix_['Y'+str(I)+','+str(T)]])
                valA = np.append(valA,float(1.0))
                irA  = np.append(irA,[ig])
                icA  = np.append(icA,[ix_['Y'+str(I)+','+str(int(v_['T-1']))]])
                valA = np.append(valA,float(-1.0))
                irA  = np.append(irA,[ig])
                icA  = np.append(icA,[ix_['Y'+str(int(v_['I+3']))+','+str(int(v_['T-1']))]])
                valA = np.append(valA,float(v_['-K/2']))
                irA  = np.append(irA,[ig])
                icA  = np.append(icA,[ix_['Y'+str(int(v_['I+3']))+','+str(T)]])
                valA = np.append(valA,float(v_['-K/2']))
        for I in range(int(v_['1']),int(v_['3'])+1):
            v_['I+3'] = 3+I
            for T in range(int(v_['1']),int(v_['N'])+1):
                v_['T-1'] = -1+T
                [ig,ig_,_] = s2mpj_ii('S'+str(int(v_['I+3']))+','+str(T),ig_)
                gtype = arrset(gtype,ig,'==')
                cnames = arrset(cnames,ig,'S'+str(int(v_['I+3']))+','+str(T))
                irA  = np.append(irA,[ig])
                icA  = np.append(icA,[ix_['Y'+str(int(v_['I+3']))+','+str(T)]])
                valA = np.append(valA,float(1.0))
                irA  = np.append(irA,[ig])
                icA  = np.append(icA,[ix_['Y'+str(int(v_['I+3']))+','+str(int(v_['T-1']))]])
                valA = np.append(valA,float(-1.0))
                [ig,ig_,_] = s2mpj_ii('S'+str(int(v_['I+3']))+','+str(T),ig_)
                gtype = arrset(gtype,ig,'==')
                cnames = arrset(cnames,ig,'S'+str(int(v_['I+3']))+','+str(T))
                irA  = np.append(irA,[ig])
                icA  = np.append(icA,[ix_['U'+str(I)+','+str(int(v_['T-1']))]])
                valA = np.append(valA,float(v_['-K/2']))
                [ig,ig_,_] = s2mpj_ii('S'+str(int(v_['I+3']))+','+str(T),ig_)
                gtype = arrset(gtype,ig,'==')
                cnames = arrset(cnames,ig,'S'+str(int(v_['I+3']))+','+str(T))
                irA  = np.append(irA,[ig])
                icA  = np.append(icA,[ix_['U'+str(I)+','+str(T)]])
                valA = np.append(valA,float(v_['-K/2']))
        for T in range(int(v_['1']),int(v_['N'])+1):
            v_['T-1'] = -1+T
            [ig,ig_,_] = s2mpj_ii('S'+str(int(v_['7']))+','+str(T),ig_)
            gtype = arrset(gtype,ig,'==')
            cnames = arrset(cnames,ig,'S'+str(int(v_['7']))+','+str(T))
            irA  = np.append(irA,[ig])
            icA  = np.append(icA,[ix_['Y'+str(int(v_['7']))+','+str(T)]])
            valA = np.append(valA,float(1.0))
            irA  = np.append(irA,[ig])
            icA  = np.append(icA,[ix_['Y'+str(int(v_['7']))+','+str(int(v_['T-1']))]])
            valA = np.append(valA,float(-1.0))
        #%%%%%%%%%%%%%% GLOBAL DIMENSIONS %%%%%%%%%%%%%%%%%
        self.n   = len(ix_)
        ngrp   = len(ig_)
        legrps = np.where(gtype=='<=')[0]
        eqgrps = np.where(gtype=='==')[0]
        gegrps = np.where(gtype=='>=')[0]
        self.nle = len(legrps)
        self.neq = len(eqgrps)
        self.nge = len(gegrps)
        self.m   = self.nle+self.neq+self.nge
        self.congrps = np.concatenate((legrps,eqgrps,gegrps))
        self.cnames = cnames[self.congrps]
        self.nob = ngrp-self.m
        self.objgrps = np.where(gtype=='<>')[0]
        #%%%%%%%%%%%%%%%%%%%  BOUNDS %%%%%%%%%%%%%%%%%%%%%
        self.xlower = np.full((self.n,1),-float('Inf'))
        self.xupper = np.full((self.n,1),+float('Inf'))
        for I in range(int(v_['1']),int(v_['3'])+1):
            for T in range(int(v_['0']),int(v_['N'])+1):
                self.xlower[ix_['U'+str(I)+','+str(T)]] = -1.0
                self.xupper[ix_['U'+str(I)+','+str(T)]] = 1.0
        self.xlower[ix_['Y'+str(int(v_['1']))+','+str(int(v_['0']))]] = 1000.0
        self.xupper[ix_['Y'+str(int(v_['1']))+','+str(int(v_['0']))]] = 1000.0
        self.xlower[ix_['Y'+str(int(v_['2']))+','+str(int(v_['0']))]] = 1000.0
        self.xupper[ix_['Y'+str(int(v_['2']))+','+str(int(v_['0']))]] = 1000.0
        self.xlower[ix_['Y'+str(int(v_['3']))+','+str(int(v_['0']))]] = 1000.0
        self.xupper[ix_['Y'+str(int(v_['3']))+','+str(int(v_['0']))]] = 1000.0
        self.xlower[ix_['Y'+str(int(v_['4']))+','+str(int(v_['0']))]] = -10.0
        self.xupper[ix_['Y'+str(int(v_['4']))+','+str(int(v_['0']))]] = -10.0
        self.xlower[ix_['Y'+str(int(v_['5']))+','+str(int(v_['0']))]] = 10.0
        self.xupper[ix_['Y'+str(int(v_['5']))+','+str(int(v_['0']))]] = 10.0
        self.xlower[ix_['Y'+str(int(v_['6']))+','+str(int(v_['0']))]] = -10.0
        self.xupper[ix_['Y'+str(int(v_['6']))+','+str(int(v_['0']))]] = -10.0
        self.xlower[ix_['Y'+str(int(v_['7']))+','+str(int(v_['0']))]] = 0.0
        self.xupper[ix_['Y'+str(int(v_['7']))+','+str(int(v_['0']))]] = 0.0
        for I in range(int(v_['1']),int(v_['6'])+1):
            self.xlower[ix_['Y'+str(I)+','+str(int(v_['N']))]] = 0.0
            self.xupper[ix_['Y'+str(I)+','+str(int(v_['N']))]] = 0.0
        #%%%%%%%%%%%%%%%%%%% START POINT %%%%%%%%%%%%%%%%%%
        self.x0 = np.zeros((self.n,1))
        self.y0 = np.zeros((self.m,1))
        #%%%%%%%%%%%%%%%%%%%% ELFTYPE %%%%%%%%%%%%%%%%%%%%%
        iet_  = {}
        elftv = []
        [it,iet_,_] = s2mpj_ii( 'eSQ', iet_)
        elftv = loaset(elftv,it,0,'V')
        #%%%%%%%%%%%%%%%%%% ELEMENT USES %%%%%%%%%%%%%%%%%%
        ie_ = {}
        self.elftype = np.array([])
        ielftype     = np.array([])
        self.elvar   = []
        for T in range(int(v_['0']),int(v_['N'])+1):
            for I in range(int(v_['1']),int(v_['3'])+1):
                ename = 'E'+str(I)+','+str(T)
                [ie,ie_,_] = s2mpj_ii(ename,ie_)
                self.elftype = arrset(self.elftype,ie,'eSQ')
                ielftype = arrset(ielftype,ie,iet_["eSQ"])
                vname = 'U'+str(I)+','+str(T)
                [iv,ix_] = s2mpj_nlx(self,vname,ix_,1,float('-Inf'),float('Inf'),None)
                posev = np.where(elftv[ielftype[ie]]=='V')[0]
                self.elvar = loaset(self.elvar,ie,posev[0],iv)
        #%%%%%%%%%%%%%%%%%%% GROUP USES %%%%%%%%%%%%%%%%%%%
        self.grelt   = []
        for ig in np.arange(0,ngrp):
            self.grelt.append(np.array([]))
        self.grftype = np.array([])
        self.grelw   = []
        nlc         = np.array([])
        for T in range(int(v_['1']),int(v_['N'])+1):
            v_['T-1'] = -1+T
            for I in range(int(v_['1']),int(v_['3'])+1):
                ig = ig_['S'+str(int(v_['7']))+','+str(T)]
                posel = len(self.grelt[ig])
                self.grelt  = (
                      loaset(self.grelt,ig,posel,ie_['E'+str(I)+','+str(int(v_['T-1']))]))
                nlc = np.union1d(nlc,np.array([ig]))
                self.grelw = loaset(self.grelw,ig,posel,float(v_['-K/2']))
                posel = len(self.grelt[ig])
                self.grelt = loaset(self.grelt,ig,posel,ie_['E'+str(I)+','+str(T)])
                nlc = np.union1d(nlc,np.array([ig]))
                self.grelw = loaset(self.grelw,ig,posel,float(v_['-K/2']))
        #%%%%%%%%%%%%%%%%%% OBJECT BOUNDS %%%%%%%%%%%%%%%%%
        self.objlower = 0.0
#    Solution
# LO SOLTN(10)           1.14735202967
# LO SOLTN(100)          1.11631518169
# LO SOLTN(1000)         1.11598643493
# LO SOLTN(2000)         1.11587382445
        #%%%%%%%% BUILD THE SPARSE MATRICES %%%%%%%%%%%%%%%
        self.A = csr_matrix((valA,(irA,icA)),shape=(ngrp,self.n))
        #%%%%%%%% DEFAULT FOR MISSING SECTION(S) %%%%%%%%%%
        #%%%%%%%%%%%%% FORM clower AND cupper %%%%%%%%%%%%%
        self.clower = np.full((self.m,1),-float('Inf'))
        self.cupper = np.full((self.m,1),+float('Inf'))
        self.clower[np.arange(self.nle,self.nle+self.neq)] = np.zeros((self.neq,1))
        self.cupper[np.arange(self.nle,self.nle+self.neq)] = np.zeros((self.neq,1))
        #%%%% RETURN VALUES FROM THE __INIT__ METHOD %%%%%%
        self.lincons  = (
              np.where(np.isin(self.congrps,np.setdiff1d(self.congrps,nlc)))[0])
        self.pbclass   = "C-CLQR2-MN-V-V"
        self.x0        = np.zeros((self.n,1))
        self.objderlvl = 2
        self.conderlvl = [2]

# **********************
#  SET UP THE FUNCTION *
#  AND RANGE ROUTINES  *
# **********************

    #%%%%%%%%%%%%%%% NONLINEAR ELEMENTS %%%%%%%%%%%%%%%

    @staticmethod
    def eSQ(self, nargout,*args):

        import numpy as np
        EV_  = args[0]
        iel_ = args[1]
        f_   = EV_[0]*EV_[0]
        if not isinstance( f_, float ):
            f_   = f_.item();
        if nargout>1:
            try:
                dim = len(IV_)
            except:
                dim = len(EV_)
            g_ = np.zeros(dim)
            g_[0] = EV_[0]+EV_[0]
            if nargout>2:
                H_ = np.zeros((1,1))
                H_[0,0] = 2.0
        if nargout == 1:
            return f_
        elif nargout == 2:
            return f_,g_
        elif nargout == 3:
            return f_,g_,H_

#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

File: julia_problems/UBH5.jl
function UBH5(action::String,args::Union{PBM,Int,Float64,Vector{Int},Vector{Float64}}...)
# 
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
# 
# 
#    Problem : UBH5
#    *********
# 
#    The problem is to minimize the integral of the control magnitude needed
#    to bring a vehicle, from given position and velocity, to the origin with
#    zero velocity in a fixed amount of time.  The controls are the components
#    of the vehicle acceleration. The discretization uses the trapezoidal rule.
#    This version of the problem is a variant of UBH1, where the cumulative
#    value of the objective is maintained as an additional state variable.
# 
#    The problem is convex.
# 
#    Source: unscaled problem 5 
#    (ODE = 1, CLS = 2, GRD = 1, MET = T, SEED = 0.) in
#    J.T. Betts and W.P. Huffman,
#    "Sparse Nonlinear Programming Test Problems (Release 1.0)",
#    Boeing Computer services, Seattle, July 1993.
# 
#    SIF input: Ph.L. Toint, October 1993.
# 
#    classification = "C-LQR2-MN-V-V"
# 
#    Number of grid points
# 
#       Alternative values for the SIF file parameters:
# IE N                   10             $-PARAMETER n=100, m=70    original value
# IE N                   100            $-PARAMETER n=1000, m=700
# IE N                   500            $-PARAMETER n=5000, m=3500
# 
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#   Translated to Julia by S2MPJ version 7 X 2024
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    name = "UBH5"

    if action == "setup"
        pb           = PB(name)
        pbm          = PBM(name)
        nargin       = length(args)
        pbm.call     = getfield( Main, Symbol( name ) )

        #%%%%%%%%%%%%%%%%%%%  PREAMBLE %%%%%%%%%%%%%%%%%%%%
        v_  = Dict{String,Float64}();
        ix_ = Dict{String,Int}();
        ig_ = Dict{String,Int}();
        if nargin<1
            v_["N"] = Int64(10);  #  SIF file default value
        else
            v_["N"] = Int64(args[1]);
        end
# IE N                   1000           $-PARAMETER n=10000, m=7000
# IE N                   2000           $-PARAMETER n=20000, m=14000
        v_["T0"] = 0.0
        v_["TF"] = 1000.0
        v_["RN"] = Float64(v_["N"])
        v_["TTIME"] = v_["TF"]-v_["T0"]
        v_["K"] = v_["TTIME"]/v_["RN"]
        v_["-K/2"] = -0.5*v_["K"]
        v_["0"] = 0
        v_["1"] = 1
        v_["2"] = 2
        v_["3"] = 3
        v_["4"] = 4
        v_["5"] = 5
        v_["6"] = 6
        v_["7"] = 7
        #%%%%%%%%%%%%%%%%%%%  VARIABLES %%%%%%%%%%%%%%%%%%%%
        pb.xscale = Float64[]
        intvars = Int64[]
        binvars = Int64[]
        for I = Int64(v_["1"]):Int64(v_["7"])
            for T = Int64(v_["0"]):Int64(v_["N"])
                iv,ix_,_ = s2mpj_ii("Y"*string(I)*","*string(T),ix_)
                arrset(pb.xnames,iv,"Y"*string(I)*","*string(T))
            end
        end
        for I = Int64(v_["1"]):Int64(v_["3"])
            for T = Int64(v_["0"]):Int64(v_["N"])
                iv,ix_,_ = s2mpj_ii("U"*string(I)*","*string(T),ix_)
                arrset(pb.xnames,iv,"U"*string(I)*","*string(T))
            end
        end
        #%%%%%%%%%%%%%%%%%%  DATA GROUPS %%%%%%%%%%%%%%%%%%%
        gtype    = String[]
        ig,ig_,_ = s2mpj_ii("OBJ",ig_)
        arrset(gtype,ig,"<>")
        iv = ix_["Y"*string(Int64(v_["7"]))*","*string(Int64(v_["N"]))]
        pbm.A[ig,iv] += Float64(1.0)
        for I = Int64(v_["1"]):Int64(v_["3"])
            v_["I+3"] = 3+I
            for T = Int64(v_["1"]):Int64(v_["N"])
                v_["T-1"] = -1+T
                ig,ig_,_ = s2mpj_ii("S"*string(I)*","*string(T),ig_)
                arrset(gtype,ig,"==")
                arrset(pb.cnames,ig,"S"*string(I)*","*string(T))
                iv = ix_["Y"*string(I)*","*string(T)]
                pbm.A[ig,iv] += Float64(1.0)
                iv = ix_["Y"*string(I)*","*string(Int64(v_["T-1"]))]
                pbm.A[ig,iv] += Float64(-1.0)
                iv = ix_["Y"*string(Int64(v_["I+3"]))*","*string(Int64(v_["T-1"]))]
                pbm.A[ig,iv] += Float64(v_["-K/2"])
                iv = ix_["Y"*string(Int64(v_["I+3"]))*","*string(T)]
                pbm.A[ig,iv] += Float64(v_["-K/2"])
            end
        end
        for I = Int64(v_["1"]):Int64(v_["3"])
            v_["I+3"] = 3+I
            for T = Int64(v_["1"]):Int64(v_["N"])
                v_["T-1"] = -1+T
                ig,ig_,_ = s2mpj_ii("S"*string(Int64(v_["I+3"]))*","*string(T),ig_)
                arrset(gtype,ig,"==")
                arrset(pb.cnames,ig,"S"*string(Int64(v_["I+3"]))*","*string(T))
                iv = ix_["Y"*string(Int64(v_["I+3"]))*","*string(T)]
                pbm.A[ig,iv] += Float64(1.0)
                iv = ix_["Y"*string(Int64(v_["I+3"]))*","*string(Int64(v_["T-1"]))]
                pbm.A[ig,iv] += Float64(-1.0)
                ig,ig_,_ = s2mpj_ii("S"*string(Int64(v_["I+3"]))*","*string(T),ig_)
                arrset(gtype,ig,"==")
                arrset(pb.cnames,ig,"S"*string(Int64(v_["I+3"]))*","*string(T))
                iv = ix_["U"*string(I)*","*string(Int64(v_["T-1"]))]
                pbm.A[ig,iv] += Float64(v_["-K/2"])
                ig,ig_,_ = s2mpj_ii("S"*string(Int64(v_["I+3"]))*","*string(T),ig_)
                arrset(gtype,ig,"==")
                arrset(pb.cnames,ig,"S"*string(Int64(v_["I+3"]))*","*string(T))
                iv = ix_["U"*string(I)*","*string(T)]
                pbm.A[ig,iv] += Float64(v_["-K/2"])
            end
        end
        for T = Int64(v_["1"]):Int64(v_["N"])
            v_["T-1"] = -1+T
            ig,ig_,_ = s2mpj_ii("S"*string(Int64(v_["7"]))*","*string(T),ig_)
            arrset(gtype,ig,"==")
            arrset(pb.cnames,ig,"S"*string(Int64(v_["7"]))*","*string(T))
            iv = ix_["Y"*string(Int64(v_["7"]))*","*string(T)]
            pbm.A[ig,iv] += Float64(1.0)
            iv = ix_["Y"*string(Int64(v_["7"]))*","*string(Int64(v_["T-1"]))]
            pbm.A[ig,iv] += Float64(-1.0)
        end
        #%%%%%%%%%%%%%% GLOBAL DIMENSIONS %%%%%%%%%%%%%%%%%
        pb.n   = length(ix_)
        ngrp   = length(ig_)
        legrps = findall(x->x=="<=",gtype)
        eqgrps = findall(x->x=="==",gtype)
        gegrps = findall(x->x==">=",gtype)
        pb.nle = length(legrps)
        pb.neq = length(eqgrps)
        pb.nge = length(gegrps)
        pb.m   = pb.nle+pb.neq+pb.nge
        pbm.congrps = [[legrps;eqgrps];gegrps]
        pb.nob = ngrp-pb.m
        pbm.objgrps = findall(x->x=="<>",gtype)
        #%%%%%%%%%%%%%%%%%%%  BOUNDS %%%%%%%%%%%%%%%%%%%%%
        pb.xlower = -1*fill(Inf,pb.n)
        pb.xupper =    fill(Inf,pb.n)
        pb.xlower = zeros(Float64,pb.n)
        for I = Int64(v_["1"]):Int64(v_["3"])
            for T = Int64(v_["0"]):Int64(v_["N"])
                pb.xlower[ix_["U"*string(I)*","*string(T)]] = -1.0
                pb.xupper[ix_["U"*string(I)*","*string(T)]] = 1.0
            end
        end
        pb.xlower[ix_["Y"*string(Int64(v_["1"]))*","*string(Int64(v_["0"]))]]  = (
              1000.0)
        pb.xupper[ix_["Y"*string(Int64(v_["1"]))*","*string(Int64(v_["0"]))]]  = (
              1000.0)
        pb.xlower[ix_["Y"*string(Int64(v_["2"]))*","*string(Int64(v_["0"]))]]  = (
              1000.0)
        pb.xupper[ix_["Y"*string(Int64(v_["2"]))*","*string(Int64(v_["0"]))]]  = (
              1000.0)
        pb.xlower[ix_["Y"*string(Int64(v_["3"]))*","*string(Int64(v_["0"]))]]  = (
              1000.0)
        pb.xupper[ix_["Y"*string(Int64(v_["3"]))*","*string(Int64(v_["0"]))]]  = (
              1000.0)
        pb.xlower[ix_["Y"*string(Int64(v_["4"]))*","*string(Int64(v_["0"]))]] = (-
             10.0)
        pb.xupper[ix_["Y"*string(Int64(v_["4"]))*","*string(Int64(v_["0"]))]] = (-
             10.0)
        pb.xlower[ix_["Y"*string(Int64(v_["5"]))*","*string(Int64(v_["0"]))]] = 10.0
        pb.xupper[ix_["Y"*string(Int64(v_["5"]))*","*string(Int64(v_["0"]))]] = 10.0
        pb.xlower[ix_["Y"*string(Int64(v_["6"]))*","*string(Int64(v_["0"]))]] = (-
             10.0)
        pb.xupper[ix_["Y"*string(Int64(v_["6"]))*","*string(Int64(v_["0"]))]] = (-
             10.0)
        pb.xlower[ix_["Y"*string(Int64(v_["7"]))*","*string(Int64(v_["0"]))]] = 0.0
        pb.xupper[ix_["Y"*string(Int64(v_["7"]))*","*string(Int64(v_["0"]))]] = 0.0
        for I = Int64(v_["1"]):Int64(v_["6"])
            pb.xlower[ix_["Y"*string(I)*","*string(Int64(v_["N"]))]] = 0.0
            pb.xupper[ix_["Y"*string(I)*","*string(Int64(v_["N"]))]] = 0.0
        end
        #%%%%%%%%%%%%%%%%%%% START POINT %%%%%%%%%%%%%%%%%%
        pb.x0 = zeros(Float64,pb.n)
        pb.y0 = zeros(Float64,pb.m)
        #%%%%%%%%%%%%%%%%%%%% ELFTYPE %%%%%%%%%%%%%%%%%%%%%
        iet_  = Dict{String,Int}()
        elftv = Vector{Vector{String}}()
        it,iet_,_ = s2mpj_ii( "eSQ", iet_)
        loaset(elftv,it,1,"V")
        #%%%%%%%%%%%%%%%%%% ELEMENT USES %%%%%%%%%%%%%%%%%%
        ie_      = Dict{String,Int}()
        ielftype = Vector{Int64}()
        for T = Int64(v_["0"]):Int64(v_["N"])
            for I = Int64(v_["1"]):Int64(v_["3"])
                ename = "E"*string(I)*","*string(T)
                ie,ie_,_  = s2mpj_ii(ename,ie_)
                arrset(pbm.elftype,ie,"eSQ")
                arrset(ielftype,ie,iet_["eSQ"])
                vname = "U"*string(I)*","*string(T)
                iv,ix_,pb = s2mpj_nlx(vname,ix_,pb,1,nothing,nothing,nothing)
                posev = findfirst(x->x=="V",elftv[ielftype[ie]])
                loaset(pbm.elvar,ie,posev,iv)
            end
        end
        #%%%%%%%%%%%%%%%%%%% GROUP USES %%%%%%%%%%%%%%%%%%%
        for ig in 1:ngrp
            arrset(pbm.grelt,ig,Int64[])
        end
        nlc = Int64[]
        for T = Int64(v_["1"]):Int64(v_["N"])
            v_["T-1"] = -1+T
            for I = Int64(v_["1"]):Int64(v_["3"])
                ig = ig_["S"*string(Int64(v_["7"]))*","*string(T)]
                posel = length(pbm.grelt[ig])+1
                loaset(pbm.grelt,ig,posel,ie_["E"*string(I)*","*string(Int64(v_["T-1"]))])
                arrset(nlc,length(nlc)+1,ig)
                loaset(pbm.grelw,ig,posel,Float64(v_["-K/2"]))
                posel = length(pbm.grelt[ig])+1
                loaset(pbm.grelt,ig,posel,ie_["E"*string(I)*","*string(T)])
                arrset(nlc,length(nlc)+1,ig)
                loaset(pbm.grelw,ig,posel,Float64(v_["-K/2"]))
            end
        end
        #%%%%%%%%%%%%%%%%%% OBJECT BOUNDS %%%%%%%%%%%%%%%%%
        pb.objlower = 0.0
#    Solution
# LO SOLTN(10)           1.14735202967
# LO SOLTN(100)          1.11631518169
# LO SOLTN(1000)         1.11598643493
# LO SOLTN(2000)         1.11587382445
        #%%%%%%%% DEFAULT FOR MISSING SECTION(S) %%%%%%%%%%
        #%%%%%%%%%%%%% FORM clower AND cupper %%%%%%%%%%%%%
        pb.clower = -1*fill(Inf,pb.m)
        pb.cupper =    fill(Inf,pb.m)
        pb.clower[pb.nle+1:pb.nle+pb.neq] = zeros(Float64,pb.neq)
        pb.cupper[pb.nle+1:pb.nle+pb.neq] = zeros(Float64,pb.neq)
        Asave = pbm.A[1:ngrp, 1:pb.n]
        pbm.A = Asave
        pbm.H = spzeros(Float64,0,0)
        #%%%%% RETURN VALUES FROM THE SETUP ACTION %%%%%%%%
        pb.lincons = findall(x-> x in setdiff( pbm.congrps,nlc),pbm.congrps)
        pb.pbclass = "C-LQR2-MN-V-V"
        pb.x0          = zeros(Float64,pb.n)
        pbm.objderlvl = 2
        pb.objderlvl = pbm.objderlvl;
        pbm.conderlvl = [2]
        pb.conderlvl  = pbm.conderlvl;
        return pb, pbm

# **********************
#  SET UP THE FUNCTION *
#  AND RANGE ROUTINES  *
# **********************

    #%%%%%%%%%%%%%%% NONLINEAR ELEMENTS %%%%%%%%%%%%%%%

    elseif action == "eSQ"

        EV_     = args[1]
        iel_    = args[2]
        nargout = args[3]
        pbm     = args[4]
        f_   = EV_[1]*EV_[1]
        if nargout>1
            dim = try length(IV_) catch; length(EV_) end
            g_  = zeros(Float64,dim)
            g_[1] = EV_[1]+EV_[1]
            if nargout>2
                H_ = zeros(Float64,1,1)
                H_[1,1] = 2.0
            end
        end
        if nargout == 1
            return f_
        elseif nargout == 2
            return f_,g_
        elseif nargout == 3
            return f_,g_,H_
        end

    #%%%%%%%%%%%%%%% THE MAIN ACTIONS %%%%%%%%%%%%%%%

    elseif action in  ["fx","fgx","fgHx","cx","cJx","cJHx","cIx","cIJx","cIJHx","cIJxv","fHxv",
                       "cJxv","cJtxv","cIJtxv","Lxy","Lgxy","LgHxy","LIxy","LIgxy","LIgHxy",
                       "LHxyv","LIHxyv"]

        pbm = args[1]
        if pbm.name == name
            pbm.has_globs = [0,0]
            return s2mpj_eval(action,args...)
        else
            println("ERROR: please run "*name*" with action = setup")
            return ntuple(i->undef,args[end])
        end

    else
        println("ERROR: action "*action*" unavailable for problem "*name*".jl")
        return ntuple(i->undef,args[end])
    end

end

#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

